File: test/instr/RemoveUnusedArgsTest.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <gtest/gtest.h>

#include "Resolver.h"
#include "Show.h"
#include "VerifyUtil.h"

namespace {

// Checks only the first callsite in method against num_args_expected
void check_callsite_regs(DexMethod* method, int num_args_expected) {
  for (const auto& mie : InstructionIterable(method->get_code())) {
    auto insn = mie.insn;
    if (opcode::is_an_invoke(insn->opcode())) {
      auto actual_method = insn->get_method();
      EXPECT_EQ(insn->srcs_size(), num_args_expected) << show(actual_method);
      break;
    }
  }
}

// Checks only the first return, whether it returns a value
void check_return(DexMethod* method, bool value) {
  for (const auto& mie : InstructionIterable(method->get_code())) {
    auto insn = mie.insn;
    if (opcode::is_a_return(insn->opcode())) {
      EXPECT_EQ(opcode::is_a_return_value(insn->opcode()), value);
      break;
    }
  }
}

// Finds vmethod with particular name and proto
DexMethod* find_vmethod(const DexClass& cls,
                        const char* name,
                        const char* proto) {
  auto vmethods = cls.get_vmethods();
  fprintf(stderr, "===\n");
  for (auto m : vmethods) {
    fprintf(stderr, "%s %s\n", SHOW(m->get_name()), SHOW(m->get_proto()));
  }
  auto it = std::find_if(vmethods.begin(), vmethods.end(),
                         [name, proto](DexMethod* m) {
                           return strcmp(name, m->get_name()->c_str()) == 0 &&
                                  strcmp(proto, SHOW(m->get_proto())) == 0;
                         });
  return it == vmethods.end() ? nullptr : *it;
}

// Sanity check: three foo constructors are defined
TEST_F(PreVerify, CtorsDefined) {
  auto foo = find_class_named(classes, "Lcom/facebook/redex/test/instr/Foo;");
  ASSERT_NE(nullptr, foo);

  auto ctors = foo->get_ctors();
  EXPECT_EQ(ctors.size(), 3);
  for (size_t i = 0; i < 3; ++i) {
    auto ctor = ctors.at(i);
    ASSERT_NE(nullptr, ctor);
  }
}

// Check unused arguments are successfully removed in constructors
TEST_F(PreVerify, RemoveCtorArg) {
  auto foo_user =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/FooUser;");
  ASSERT_NE(nullptr, foo_user);

  auto use_foo = find_vmethod_named(*foo_user, "use_foo1");
  ASSERT_NE(nullptr, use_foo);
  use_foo->balloon();

  check_callsite_regs(use_foo, 2);
}

TEST_F(PostVerify, RemoveCtorArg) {
  auto foo_user =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/FooUser;");
  ASSERT_NE(nullptr, foo_user);

  auto use_foo = find_vmethod_named(*foo_user, "use_foo1");
  ASSERT_NE(nullptr, use_foo);
  use_foo->balloon();

  check_callsite_regs(use_foo, 1);
}

// Check arguments of constructor aren't removed when they're used
TEST_F(PreVerify, DontRemoveUsedCtorArg) {
  auto foo_user =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/FooUser;");
  ASSERT_NE(nullptr, foo_user);

  auto use_foo2 = find_vmethod_named(*foo_user, "use_foo2");
  ASSERT_NE(nullptr, use_foo2);
  use_foo2->balloon();

  check_callsite_regs(use_foo2, 3);
}

TEST_F(PostVerify, DontRemoveUsedCtorArg) {
  auto foo_user =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/FooUser;");
  ASSERT_NE(nullptr, foo_user);

  auto use_foo2 = find_vmethod_named(*foo_user, "use_foo2");
  ASSERT_NE(nullptr, use_foo2);
  use_foo2->balloon();

  check_callsite_regs(use_foo2, 3);
}

// Check arguments of constructor aren't removed when the signature collides
TEST_F(PreVerify, CollidingCtorArg) {
  auto foo_user =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/FooUser;");
  ASSERT_NE(nullptr, foo_user);

  auto use_foo3 = find_vmethod_named(*foo_user, "use_foo3");
  ASSERT_NE(nullptr, use_foo3);
  use_foo3->balloon();

  check_callsite_regs(use_foo3, 4);
}

TEST_F(PostVerify, CollidingCtorArg) {
  auto foo_user =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/FooUser;");
  ASSERT_NE(nullptr, foo_user);

  auto use_foo3 = find_vmethod_named(*foo_user, "use_foo3");
  ASSERT_NE(nullptr, use_foo3);
  use_foo3->balloon();

  check_callsite_regs(use_foo3, 4);
}

// Check no-argument static methods' invokes don't change
TEST_F(PreVerify, StaticsNoArgs) {
  auto statics =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/Statics;");
  ASSERT_NE(nullptr, statics);

  auto static1 = find_dmethod_named(*statics, "static1");
  ASSERT_NE(nullptr, static1);

  auto statics_user =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/StaticsUser;");
  ASSERT_NE(nullptr, statics_user);

  auto use_static1 = find_vmethod_named(*statics_user, "use_static1");
  ASSERT_NE(nullptr, use_static1);
  use_static1->balloon();

  check_callsite_regs(use_static1, 0);
}

TEST_F(PostVerify, StaticsNoArgs) {
  auto statics_user =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/StaticsUser;");
  ASSERT_NE(nullptr, statics_user);

  auto use_static1 = find_vmethod_named(*statics_user, "use_static1");
  ASSERT_NE(nullptr, use_static1);
  use_static1->balloon();

  check_callsite_regs(use_static1, 0);
}

// Check static methods' invokes don't change when args are used
TEST_F(PreVerify, StaticsUsedArgs) {
  auto statics =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/Statics;");
  ASSERT_NE(nullptr, statics);

  auto static2 = find_dmethod_named(*statics, "static2");
  ASSERT_NE(nullptr, static2);

  auto statics_user =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/StaticsUser;");
  ASSERT_NE(nullptr, statics_user);

  auto use_static2 = find_vmethod_named(*statics_user, "use_static2");
  ASSERT_NE(nullptr, use_static2);
  use_static2->balloon();

  check_callsite_regs(use_static2, 1);
}

TEST_F(PostVerify, StaticsUsedArgs) {
  auto statics_user =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/StaticsUser;");
  ASSERT_NE(nullptr, statics_user);

  auto use_static2 = find_vmethod_named(*statics_user, "use_static2");
  ASSERT_NE(nullptr, use_static2);
  use_static2->balloon();

  check_callsite_regs(use_static2, 1);
}

// Check static method arg removal for unused args
TEST_F(PreVerify, StaticsUnusedArgs) {
  auto statics =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/Statics;");
  ASSERT_NE(nullptr, statics);

  auto static3 = find_dmethod_named(*statics, "static3");
  ASSERT_NE(nullptr, static3);

  auto statics_user =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/StaticsUser;");
  ASSERT_NE(nullptr, statics_user);

  auto use_static3 = find_vmethod_named(*statics_user, "use_static3");
  ASSERT_NE(nullptr, use_static3);
  use_static3->balloon();

  check_callsite_regs(use_static3, 2);
}

TEST_F(PostVerify, StaticsUnusedArgs) {
  auto statics_user =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/StaticsUser;");
  ASSERT_NE(nullptr, statics_user);

  auto use_static3 = find_vmethod_named(*statics_user, "use_static3$uva1$0");
  ASSERT_NE(nullptr, use_static3);
  use_static3->balloon();

  check_callsite_regs(use_static3, 1);
}

// Checks that static method result type doesn't change when result is used
TEST_F(PreVerify, StaticsUsedResult) {
  auto statics =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/Statics;");
  ASSERT_NE(nullptr, statics);

  auto static4 = find_dmethod_named(*statics, "static4_with_result");
  ASSERT_NE(nullptr, static4);

  ASSERT_FALSE(static4->get_proto()->is_void());
  static4->balloon();
  check_return(static4, true /* value */);
}

TEST_F(PostVerify, StaticsUsedResult) {
  auto statics =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/Statics;");
  ASSERT_NE(nullptr, statics);

  auto static4 = find_dmethod_named(*statics, "static4_with_result");
  ASSERT_NE(nullptr, static4);

  ASSERT_FALSE(static4->get_proto()->is_void());
  static4->balloon();
  check_return(static4, true /* value */);
}

// Check static method result removal for unused results
TEST_F(PreVerify, StaticsUnusedResult) {
  auto statics =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/Statics;");
  ASSERT_NE(nullptr, statics);

  auto static5 = find_dmethod_named(*statics, "static5_with_result");
  ASSERT_NE(nullptr, static5);

  ASSERT_FALSE(static5->get_proto()->is_void());
  static5->balloon();
  check_return(static5, true /* value */);
}

TEST_F(PostVerify, StaticsUnusedResult) {
  auto statics =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/Statics;");
  ASSERT_NE(nullptr, statics);

  auto static5 = find_dmethod_named(*statics, "static5_with_result");
  ASSERT_NE(nullptr, static5);

  ASSERT_TRUE(static5->get_proto()->is_void());
  static5->balloon();
  check_return(static5, false /* value */);
}

// Check overloaded name mangling upon collision
TEST_F(PreVerify, PrivatesUsedArgs) {
  auto privates =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/Privates;");
  ASSERT_NE(nullptr, privates);
  Scope scope = {privates};
  balloon_for_test(scope);

  // 1 default constructor + 2 private void private1() methods
  auto dmethods = privates->get_dmethods();
  EXPECT_EQ(dmethods.size(), 3);

  bool two_args_method = false;
  bool three_args_method = false;

  for (auto dmethod : dmethods) {
    if (!method::is_constructor(dmethod)) {
      auto num_args = dmethod->get_proto()->get_args()->size();
      if (num_args == 2) {
        two_args_method = true;
      } else {
        three_args_method = true;
      }
    }
  }

  EXPECT_TRUE(two_args_method && three_args_method);
  auto use_private_first = find_vmethod_named(*privates, "use_private_first");
  check_callsite_regs(use_private_first, 3);
  auto use_private_second = find_vmethod_named(*privates, "use_private_second");
  check_callsite_regs(use_private_second, 4);
}

TEST_F(PostVerify, PrivatesUsedArgs) {
  auto privates =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/Privates;");
  ASSERT_NE(nullptr, privates);
  Scope scope = {privates};
  balloon_for_test(scope);

  auto dmethods = privates->get_dmethods();
  EXPECT_EQ(dmethods.size(), 3);

  std::vector<DexMethod*> overloaded_methods;
  for (auto dmethod : dmethods) {
    if (!method::is_constructor(dmethod)) {
      auto num_args = dmethod->get_proto()->get_args()->size();
      EXPECT_EQ(num_args, 2);
      overloaded_methods.emplace_back(dmethod);
    }
  }

  EXPECT_EQ(overloaded_methods.size(), 2);
  auto name1 = overloaded_methods.at(0)->get_name()->str();
  auto name2 = overloaded_methods.at(1)->get_name()->str();
  ASSERT_NE(name1, name2);

  auto use_private_first = find_vmethod_named(*privates, "use_private_first");
  check_callsite_regs(use_private_first, 3);
  auto use_private_second = find_vmethod_named(*privates, "use_private_second");
  check_callsite_regs(use_private_second, 3);
}

// Check nonvirtual method arg removal for unused args
TEST_F(PreVerify, PublicNonVirtualsUnusedArgs) {
  auto non_virtuals =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/NonVirtuals;");
  ASSERT_NE(nullptr, non_virtuals);

  auto non_virtual1 = find_vmethod_named(*non_virtuals, "non_virtual1");
  ASSERT_NE(nullptr, non_virtual1);
  auto code = non_virtual1->get_code();

  auto non_virtuals_user = find_class_named(
      classes, "Lcom/facebook/redex/test/instr/NonVirtualsUser;");
  ASSERT_NE(nullptr, non_virtuals_user);

  auto use_non_virtual1 =
      find_vmethod_named(*non_virtuals_user, "use_non_virtual1");
  ASSERT_NE(nullptr, use_non_virtual1);
  use_non_virtual1->balloon();

  check_callsite_regs(use_non_virtual1, 2);
}

TEST_F(PostVerify, NonVirtualsUnusedArgs) {
  auto non_virtuals =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/NonVirtuals;");
  ASSERT_NE(nullptr, non_virtuals);

  auto non_virtual1 = find_vmethod_named(*non_virtuals, "non_virtual1$uva0$0");
  ASSERT_NE(nullptr, non_virtual1);

  auto non_virtuals_user = find_class_named(
      classes, "Lcom/facebook/redex/test/instr/NonVirtualsUser;");
  ASSERT_NE(nullptr, non_virtuals_user);

  auto use_non_virtual1 =
      find_vmethod_named(*non_virtuals_user, "use_non_virtual1");
  ASSERT_NE(nullptr, use_non_virtual1);
  use_non_virtual1->balloon();

  check_callsite_regs(use_non_virtual1, 1);
}

// Check protected method arg removal for unused args
TEST_F(PreVerify, ProtectedNonVirtualsUnusedArgs) {
  auto non_virtuals =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/NonVirtuals;");
  ASSERT_NE(nullptr, non_virtuals);

  auto non_virtual2 = find_vmethod_named(*non_virtuals, "non_virtual2");
  ASSERT_NE(nullptr, non_virtual2);

  auto non_virtuals_user = find_class_named(
      classes, "Lcom/facebook/redex/test/instr/NonVirtualsUser;");
  ASSERT_NE(nullptr, non_virtuals_user);

  auto use_non_virtual2 =
      find_vmethod_named(*non_virtuals_user, "use_non_virtual2");
  ASSERT_NE(nullptr, use_non_virtual2);
  use_non_virtual2->balloon();

  check_callsite_regs(use_non_virtual2, 2);
}

TEST_F(PostVerify, ProtectedNonVirtualsUnusedArgs) {
  auto non_virtuals =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/NonVirtuals;");
  ASSERT_NE(nullptr, non_virtuals);

  auto non_virtual2 = find_vmethod_named(*non_virtuals, "non_virtual2$uva0$0");
  ASSERT_NE(nullptr, non_virtual2);

  auto non_virtuals_user = find_class_named(
      classes, "Lcom/facebook/redex/test/instr/NonVirtualsUser;");
  ASSERT_NE(nullptr, non_virtuals_user);

  auto use_non_virtual2 =
      find_vmethod_named(*non_virtuals_user, "use_non_virtual2");
  ASSERT_NE(nullptr, use_non_virtual2);
  use_non_virtual2->balloon();

  check_callsite_regs(use_non_virtual2, 1);
}

// Check argument reordering
TEST_F(PreVerify, Reorderables) {
  auto reorderables =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/Reorderables;");
  ASSERT_NE(nullptr, reorderables);

  auto reorderable1 =
      find_vmethod(*reorderables, "reorderable1", "(ILjava/lang/Object;D)V");
  ASSERT_NE(nullptr, reorderable1);

  auto reorderable2 =
      find_vmethod(*reorderables, "reorderable2", "(DILjava/lang/Object;)V");
  ASSERT_NE(nullptr, reorderable2);

  auto reorderable2alt =
      find_vmethod(*reorderables, "reorderable2", "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable2alt);

  auto reorderable2altalt =
      find_vmethod(*reorderables, "reorderable2", "(Ljava/lang/Object;ID)V");
  ASSERT_NE(nullptr, reorderable2altalt);
}

TEST_F(PostVerify, Reorderables) {
  auto reorderables =
      find_class_named(classes, "Lcom/facebook/redex/test/instr/Reorderables;");
  ASSERT_NE(nullptr, reorderables);

  auto reorderable1 = find_vmethod(*reorderables, "reorderable1$rvp0$0",
                                   "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable1);

  auto reorderable2 =
      find_vmethod(*reorderables, "reorderable2", "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable2);

  auto reorderable2alt = find_vmethod(*reorderables, "reorderable2$rvp0$0",
                                      "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable2alt);

  auto reorderable2altalt = find_vmethod(*reorderables, "reorderable2$rvp0$1",
                                         "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable2altalt);
}

TEST_F(PreVerify, ReorderablesInterface) {
  auto reorderables = find_class_named(
      classes, "Lcom/facebook/redex/test/instr/ReorderablesInterface;");
  ASSERT_NE(nullptr, reorderables);

  auto reorderable1 =
      find_vmethod(*reorderables, "reorderable1", "(ILjava/lang/Object;D)V");
  ASSERT_NE(nullptr, reorderable1);

  auto reorderable2 =
      find_vmethod(*reorderables, "reorderable2", "(DILjava/lang/Object;)V");
  ASSERT_NE(nullptr, reorderable2);

  auto reorderable2alt =
      find_vmethod(*reorderables, "reorderable2", "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable2alt);

  auto reorderable2altalt =
      find_vmethod(*reorderables, "reorderable2", "(Ljava/lang/Object;ID)V");
  ASSERT_NE(nullptr, reorderable2altalt);
}

TEST_F(PostVerify, ReorderablesInterface) {
  auto reorderables = find_class_named(
      classes, "Lcom/facebook/redex/test/instr/ReorderablesInterface;");
  ASSERT_NE(nullptr, reorderables);

  auto reorderable1 = find_vmethod(*reorderables, "reorderable1$rvp0$0",
                                   "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable1);

  auto reorderable2 =
      find_vmethod(*reorderables, "reorderable2", "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable2);

  auto reorderable2alt = find_vmethod(*reorderables, "reorderable2$rvp0$0",
                                      "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable2alt);

  auto reorderable2altalt = find_vmethod(*reorderables, "reorderable2$rvp0$1",
                                         "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable2altalt);
}

TEST_F(PreVerify, SubReorderables) {
  auto reorderables = find_class_named(
      classes, "Lcom/facebook/redex/test/instr/SubReorderables;");
  ASSERT_NE(nullptr, reorderables);

  auto reorderable1 =
      find_vmethod(*reorderables, "reorderable1", "(ILjava/lang/Object;D)V");
  ASSERT_NE(nullptr, reorderable1);

  auto reorderable2 =
      find_vmethod(*reorderables, "reorderable2", "(DILjava/lang/Object;)V");
  ASSERT_NE(nullptr, reorderable2);

  auto reorderable2alt =
      find_vmethod(*reorderables, "reorderable2", "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable2alt);

  auto reorderable2altalt =
      find_vmethod(*reorderables, "reorderable2", "(Ljava/lang/Object;ID)V");
  ASSERT_NE(nullptr, reorderable2altalt);
}

TEST_F(PostVerify, SubReorderables) {
  auto reorderables = find_class_named(
      classes, "Lcom/facebook/redex/test/instr/SubReorderables;");
  ASSERT_NE(nullptr, reorderables);

  auto reorderable1 = find_vmethod(*reorderables, "reorderable1$rvp0$0",
                                   "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable1);

  auto reorderable2 =
      find_vmethod(*reorderables, "reorderable2", "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable2);

  auto reorderable2alt = find_vmethod(*reorderables, "reorderable2$rvp0$0",
                                      "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable2alt);

  auto reorderable2altalt = find_vmethod(*reorderables, "reorderable2$rvp0$1",
                                         "(Ljava/lang/Object;DI)V");
  ASSERT_NE(nullptr, reorderable2altalt);
}

} // namespace

File: test/instr/RemoveUnusedArgsTest.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.redex.test.instr;

import javax.security.auth.x500.X500Principal;

class Foo {
  private int x;

  // One removable arg
  public Foo(int x) {
    this.x = 4;
  }

  // No removable args
  public Foo(int x, int y) {
    this.x = x + y;
  }

  // One colliding removable arg
  public Foo(int x, int y, double z) {
    this.x = x + y + 3;
  }
}

class FooVirtual {
  private int x;

  // One removable arg
  public FooVirtual(int x) {
    this.x = 4;
  }

  // No removable args
  public FooVirtual(int x, int y) {
    this.x = x + y;
  }

  public void used_in_base(int x) {
    this.x = x;
  }

  public void used_in_derived(int x) {
    this.x = 3;
  }

  public void used_in_all(int x) {
    this.x = x;
  }

  public void used_in_none(int x) {
    this.x = 2;
  }

  public void used_in_derived_derived(int x) {
    this.x = 3;
  }

  public void used_in_sibling(int x) {
    this.x = 3;
  }
}

interface FooInterface {
  public void used_in_all(int x);
  public void used_in_none(int x);
  public void used_in_base(int x);
}

class FooDerived extends FooVirtual implements FooInterface {
  private int x;

  // One removable arg
  public FooDerived(int x) {
    super(x);
  }

  // No removable args
  public FooDerived(int x, int y) {
    super(x, y);
  }

  public void used_in_base(int x) {
    this.x = 3;
  }

  public void used_in_sibling(int x) {
    this.x = 3;
  }

  public void used_in_derived(int x) {
    this.x = x;
  }

  public void used_in_all(int x) {
    this.x = x;
  }

  public void used_in_none(int x) {
    this.x = 1;
  }
}

class FooDerivedDerived extends FooDerived {
  private int x;

  // One removable arg
  public FooDerivedDerived(int x) {
    super(x);
  }

  public void used_in_base(int x) {
    this.x = 3;
  }

  public void used_in_derived_derived(int x) {
    this.x = x;
  }

  public void used_in_all(int x) {
    this.x = x;
  }

  public void used_in_none(int x) {
    this.x = 1;
  }
}

class FooSibling extends FooVirtual {
  private int x;

  // One removable arg
  public FooSibling(int x) {
    super(x);
  }

  public void used_in_sibling(int x) {
    this.x = x;
  }

  public void used_in_all(int x) {
    this.x = x;
  }

  public void used_in_none(int x) {
    this.x = 1;
  }
}

abstract class FooAbstract {
  private int x;

  public abstract void abstract_method_used(int x);

  public abstract void abstract_method_unused(int x);
}

class FooAbstractDerived extends FooAbstract {
  private int x;

  public void abstract_method_used(int x) {
    this.x = x;
  }

  public void abstract_method_unused(int x) {
    this.x = 3;
  }
}

class FooUser {
  private int x;

  public void use_foo1() {
    Foo foo = new Foo(x);
  }

  public void use_foo2() {
    Foo foo = new Foo(1, 1);
  }

  public void use_foo3() {
    double d = 5.5;
    Foo foo = new Foo(1, 1, d);
  }
}

class FooVirtualUser {
  private int x;

  public void use_foo1_virtual() {
    FooVirtual foo = new FooVirtual(1);
  }

  public void use_foo1_derived() {
    FooDerived foo = new FooDerived(1);
  }

  public void use_foo2_virtual() {
    FooVirtual foo = new FooVirtual(1, 2);
  }

  public void use_foo2_derived() {
    FooDerived foo = new FooDerived(1, 2);
  }

  public void use_foo_used_in_base_virtual() {
    FooVirtual foo = new FooVirtual(1);
    foo.used_in_base(1);
  }

  public void use_foo_used_in_base_derived() {
    FooDerived foo = new FooDerived(1);
    foo.used_in_base(1);
  }

  public void use_foo_used_in_derived_virtual() {
    FooVirtual foo = new FooVirtual(1);
    foo.used_in_derived(1);
  }

  public void use_foo_used_in_derived_derived() {
    FooDerived foo = new FooDerived(1);
    foo.used_in_derived(1);
  }

  public void use_foo_used_in_all_virtual() {
    FooVirtual foo = new FooVirtual(1);
    foo.used_in_all(1);
  }

  public void use_foo_used_in_all_derived() {
    FooDerived foo = new FooDerived(1);
    foo.used_in_all(1);
  }

  public void use_foo_used_in_none_virtual() {
    FooVirtual foo = new FooVirtual(1);
    foo.used_in_none(1);
  }

  public void use_foo_used_in_none_derived() {
    FooDerived foo = new FooDerived(1);
    foo.used_in_none(1);
  }

  public void use_foo_used_in_base_derived_derived() {
    FooDerivedDerived foo = new FooDerivedDerived(1);
    foo.used_in_base(1);
  }

  public void use_foo_used_in_derived_derived_virtual() {
    FooVirtual foo = new FooVirtual(1);
    foo.used_in_derived_derived(1);
  }

  public void use_foo_used_in_sibling_derived() {
    FooDerived foo = new FooDerived(1);
    foo.used_in_sibling(1);
  }

  public void use_foo_used_in_none_interface() {
    FooInterface foo = new FooDerived(1);
    foo.used_in_none(1);
  }

  public void use_foo_used_in_all_interface() {
    FooInterface foo = new FooDerived(1);
    foo.used_in_all(1);
  }

  public void use_foo_used_in_base_interface() {
    FooInterface foo = new FooDerived(1);
    foo.used_in_base(1);
  }

  public void use_foo_abstract_used() {
    FooAbstract foo = new FooAbstractDerived();
    foo.abstract_method_used(1);
  }

  public void use_foo_abstract_unused() {
    FooAbstract foo = new FooAbstractDerived();
    foo.abstract_method_unused(1);
  }
}

class Statics {
  private static String greeting;
  private static double wide;
  public static void static1() {
    String hello = "I don't have arguments!";
  }

  public static void static2(boolean truthy) {
    String bonjour = "J'ai un argument";
    if (truthy) {
      bonjour = "Je use the argument";
    }
    Statics.greeting = bonjour;
  }

  public static void static3(Object obj, double wide_param) {
    String ciao = "Je ne sais pas Italian; two arguments";
    wide_param *= 2.0;
    ciao = "wide_param is alive, obj is dead";
    Statics.wide = wide_param;
  }

  public static int static4_with_result() {
    return 0;
  }

  public static int static5_with_result() {
    return 0;
  }
}

class StaticsUser {
  public void use_static1() {
    Statics.static1();
  }

  public void use_static2() {
    Statics.static2(true);
  }

  public void use_static3(Foo foo) {
    double dub = 4.0;
    Statics.static3(foo, dub);
  }

  public void use_static4_with_result(Foo foo) {
    int x = Statics.static4_with_result();
    x = x / x;
  }
}

class Privates {
  private double save;
  public void use_private_first() {
    private1(1.0, 2.0);
  }

  public void use_private_second() {
    private1(0, 1.0, 2.0);
  }

  // first
  private void private1(double x, double y) {
    this.save = x + y;
  }

  // second
  private void private1(int c, double x, double y) {
    this.save = x * y;
  }
}

class Parent {
}

class NonVirtuals extends Parent {
  public void non_virtual1(int x) {
    x = 5;
  }

  protected void non_virtual2(double x) {
    x = 5.0;
  }
}

class NonVirtualsUser {
  public void use_non_virtual1(NonVirtuals nv) {
    nv.non_virtual1(1);
  }

  public void use_non_virtual2(NonVirtuals nv) {
    nv.non_virtual2(1.0);
  }
}

class Reorderables {
  int int_value;
  Object obj_value;
  double double_value;

  public void reorderable1(int x, Object y, double z) {
    this.int_value = x;
    this.obj_value = y;
    this.double_value = z;
  }

  public void reorderable2(double x, int y, Object z) {
    this.double_value = x;
    this.int_value = y;
    this.obj_value = z;
  }

  public void reorderable2(Object x, double y, int z) {
    this.obj_value = x;
    this.double_value = y;
    this.int_value = z;
  }

  public void reorderable2(Object x, int y, double z) {
    this.obj_value = x;
    this.int_value = y;
    this.double_value = z;
  }
}

interface ReorderablesInterface {
  void reorderable1(int x, Object y, double z);
  void reorderable2(double x, int y, Object z);
  void reorderable2(Object x, double y, int z);
  void reorderable2(Object x, int y, double z);
}

class SubReorderables extends Reorderables implements ReorderablesInterface {
  public void reorderable1(int x, Object y, double z) {
  }

  public void reorderable2(double x, int y, Object z) {
  }

  public void reorderable2(Object x, double y, int z) {
  }

  public void reorderable2(Object x, int y, double z) {
  }
}
